pkg/ipam/poolallocator/allocator_test.go

```diff
@@ -0,0 +1,180 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package poolallocator
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	fakeversioned "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func newIPPoolAllocator(poolName string, initObjects []runtime.Object) *IPPoolAllocator {
+	crdClient := fakeversioned.NewSimpleClientset(initObjects...)
+	allocator, _ := NewIPPoolAllocator(poolName, crdClient)
+	return allocator
+}
+
+func TestAllocateIP(t *testing.T) {
+	poolName := "fakePool"
+	ipRange := crdv1a2.IPRange{
+		Start: "10.2.2.100",
+		End:   "10.2.2.120",
+	}
+	subnetSpec := crdv1a2.SubnetSpec{
+		Gateway:      "10.2.2.1",
+		PrefixLength: 24,
+	}
+	subnetRange := crdv1a2.SubnetIPRange{IPRange: ipRange,
+		SubnetSpec: subnetSpec}
+
+	pool := crdv1a2.IPPool{
+		ObjectMeta: metav1.ObjectMeta{Name: poolName},
+		Spec:       crdv1a2.IPPoolSpec{IPRanges: []crdv1a2.SubnetIPRange{subnetRange}},
+	}
+
+	allocator := newIPPoolAllocator(poolName, []runtime.Object{&pool})
+
+	// Allocate specific IP from the range
+	returnSpec, err := allocator.AllocateIP(net.ParseIP("10.2.2.101"), crdv1a2.IPPoolUsageStateAllocated, "fakePod")
+	assert.Equal(t, subnetSpec, returnSpec)
+	require.NoError(t, err)
+
+	// Validate IP outside the range is not allocated
+	returnSpec, err = allocator.AllocateIP(net.ParseIP("10.2.2.121"), crdv1a2.IPPoolUsageStateAllocated, "fakePod")
+	require.Error(t, err)
+
+	// Make sure IP allocated above is not allocated again
+	for _, expectedIP := range []string{"10.2.2.100", "10.2.2.102"} {
+		ip, returnSpec, err := allocator.AllocateNext(crdv1a2.IPPoolUsageStateAllocated, "fakePod")
+		require.NoError(t, err)
+		assert.Equal(t, net.ParseIP(expectedIP), ip)
+		assert.Equal(t, subnetSpec, returnSpec)
+	}
+}
+
+func TestAllocateNext(t *testing.T) {
+	poolName := "fakePool"
+	ipRange := crdv1a2.IPRange{
+		Start: "10.2.2.100",
+		End:   "10.2.2.120",
+	}
+	subnetSpec := crdv1a2.SubnetSpec{
+		Gateway:      "10.2.2.1",
+		PrefixLength: 24,
+	}
+	subnetRange := crdv1a2.SubnetIPRange{IPRange: ipRange,
+		SubnetSpec: subnetSpec}
+
+	pool := crdv1a2.IPPool{
+		ObjectMeta: metav1.ObjectMeta{Name: poolName},
+		Spec:       crdv1a2.IPPoolSpec{IPRanges: []crdv1a2.SubnetIPRange{subnetRange}},
+	}
+
+	allocator := newIPPoolAllocator(poolName, []runtime.Object{&pool})
+
+	for _, expectedIP := range []string{"10.2.2.100", "10.2.2.101"} {
+		ip, returnSpec, err := allocator.AllocateNext(crdv1a2.IPPoolUsageStateAllocated, "fakePod")
+		require.NoError(t, err)
+		assert.Equal(t, net.ParseIP(expectedIP), ip)
+		assert.Equal(t, subnetSpec, returnSpec)
+	}
+}
+
+func TestAllocateNextMultiRange(t *testing.T) {
```

Comment:
It might be helpful to also add a negative test for pool exhaustion

Reply:
Good point, I'm still working on tests, hoping to add test for the conflict scenario as well.